### PR TITLE
Feature/org install

### PR DESCRIPTION
**Pull request by Lance-Simeon**

### CODE REVIEW

File: OrgInstaller.ps1 [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+            $accessToOrgs = (Invoke-WithRetry { Invoke-RestMethod -Header $AuthenicationHeader -Uri "https://app.vssps.visualstudio.com/_apis/accounts?memberId=$currentUserId&api-version=5.1" -Method Get }).value
+            if ($accessToOrgs.accountName -contains "$Organization") {
+                Write-Information "DevOps Organization: $Organization exists and current user has access"
+            }
+            else {
+                throw "DevOps Organization: $Organization exists but the current user does not have access to $Organization"
+            }
+        }
+        else {
+            Write-Information "Creating DevOps Organization: $Organization"
```
Comment (admin-simeon) [inline OrgInstaller.ps1:114]: I don't suppose there is any way at all to make this a non-AAD connected org in the REST request?

File: OrgInstaller.ps1 [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+            $userToInviteDescriptor = ((Invoke-WithRetry { Invoke-RestMethod -Header $AuthenicationHeader -Uri "https://vssps.dev.azure.com/$Organization/_apis/graph/users?api-version=6.1-preview.1" -Method Get }).value |? principalName -eq "$SimeonUserToInviteToOrg").descriptor
+            Invoke-WithRetry { Invoke-RestMethod -Header $AuthenicationHeader -Uri "https://vssps.dev.azure.com/$Organization/_apis/graph/memberships/$userToInviteDescriptor/$projectCollectionAdminGroupDescriptor`?api-version=6.1-preview.1" -Method Put } | Out-Null
+        }
+    }
+
+    Write-Information "Getting project information"
+    Invoke-Command -ScriptBlock {
+        Write-Information "Getting project id for: $projectName"
+        # Set or get Project id
+        $projectId = ((Invoke-WithRetry { Invoke-RestMethod -Header $AuthenicationHeader -Uri "https://dev.azure.com/$Organization/_apis/projects?api-version=6.1-preview.4" -Method Get }).value |? { $_.name -eq "$ProjectName" }).id
```
Comment (admin-simeon) [inline OrgInstaller.ps1:172]: might just extract to Get-AzureDevOpsProjectId $Organization...and could cache the result (see how SimeonInstaller does this for tenant ids)...you're getting ProjectId over and over again

File: OrgInstaller.ps1 [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+                }
+            }
+        }
+    }
+"@
+        } | Out-Null
+
+        # Overview > uncheck Boards and Test Plans
+        Write-Information "Updating project settings turning off Boards and Test Plans"
+        Invoke-WithRetry { Invoke-RestMethod -Header $AuthenicationHeader -Uri "https://dev.azure.com/$Organization/_apis/FeatureManagement/FeatureStates/host/project/$projectId/ms.vss-work.agile?api-version=4.1-preview.1" -Method Patch -ContentType "application/json" -Body @"
```
Comment (admin-simeon) [inline OrgInstaller.ps1:242]: only do this if it's a new org

File: OrgInstaller.ps1 [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+            Write-Information "Renaming $ProjectName to default"
+            $repos = (Invoke-WithRetry { Invoke-RestMethod -Uri "https://dev.azure.com/$Organization/$projectId/_apis/git/repositories?api-version=6.0" -Method Get }).value
+            if ($repos.name -contains "$ProjectName") {
+                $repoId = ($repos |? { $_.name -eq "$ProjectName" }).id
+                Invoke-WithRetry { Invoke-RestMethod -Uri "https://dev.azure.com/$Organization/$projectId/_apis/git/repositories/$repoId`?api-version=5.0" -Method Patch -Body '{"name":"Default"}' } -ContentType "application/json" | Out-Null
+            }
+        }
+    }
+
+    $projectId = ((Invoke-WithRetry { Invoke-RestMethod -Header $AuthenicationHeader -Uri "https://dev.azure.com/$Organization/_apis/projects?api-version=6.1-preview.4" -Method Get }).value |? { $_.name -eq "$ProjectName" }).id
```
Comment (admin-simeon) [inline OrgInstaller.ps1:202]: why are you doing this again?

File: OrgInstaller.ps1 [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+        "contributionIds": [
+            "ms.vss-build-web.pipelines-general-settings-data-provider"
+        ],
+        "dataProviderContext": {
+            "properties": {
+                "enforceSettableVar": "false",
+                "enforceJobAuthScope": "false",
+                "enforceJobAuthScopeForReleases": "false",
+                "sourcePage": {
+                    "url": "https://dev.azure.com/$Organization/$ProjectName/_settings/settings",
```
Comment (admin-simeon) [inline OrgInstaller.ps1:225]: Call $Project to be consistent with SimeonInstaller.

**Review by admin-simeon — CHANGES_REQUESTED** [review]

File: OrgInstaller.ps1 [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+            if ($accessToOrgs.accountName -contains "$Organization") {
+                Write-Information "DevOps Organization: $Organization exists and current user has access"
+            }
+            else {
+                throw "DevOps Organization: $Organization exists but the current user does not have access to $Organization"
+            }
+        }
+        else {
+            Write-Information "Creating DevOps Organization: $Organization"
+            Invoke-WithRetry { Invoke-RestMethod -Headers $AuthenicationHeader -ContentType "application/json" -Method Post -Uri "https://aexprodcus1.vsaex.visualstudio.com/_apis/HostAcquisition/Collections?collectionName=$Organization`&preferredRegion=CUS&api-version=5.0-preview.2" -Body @"
```
Comment (admin-simeon) [inline OrgInstaller.ps1:115]: aexprodcus1 is for central us - I think you want aex.dev.azure.com to be region agnostic at least for the host

File: SimeonInstaller.ps1 [hunk 7]
```diff
@@ -512,6 +515,115 @@ CRLFOption=CRLFAlways
...
+        }
+    }
+
+    <#
+    .SYNOPSIS
+        Sets Azure DevOps project permissions
+    #>
+    function Set-AzureDevOpsAccessControlEntry {
+        param(
+            [string]$Organization,
```
Comment (admin-simeon) [inline SimeonInstaller.ps1:558]: add validation attributes to parameters like other cmdlets

File: SimeonInstaller.ps1 [hunk 8]
```diff
@@ -1657,6 +1769,358 @@ CRLFOption=CRLFAlways
...
+        }
+
+        Write-Information "Getting reporting email password"
+        $reportingEmailPw = (Get-AzureKeyVaultSecret -KeyVaultSecretUri $ReportingEmailPasswordKeyVaultUrl)
+        if (!$reportingEmailPw) {
+            throw "Unable to retrieve reporting email password please contact Simeon Support for assistance."
+        }
+
+        $token = Get-SimeonAzureDevOpsAccessToken
+        $AuthenicationHeader = @{Authorization = "Bearer $token" }
```
Comment (admin-simeon) [inline SimeonInstaller.ps1:1802]: camelCase

File: SimeonInstaller.ps1 [hunk 9]
```diff
@@ -512,6 +515,115 @@ CRLFOption=CRLFAlways
...
+        Gets the Azure DevOps project id for the given project
+    #>
+    function Get-AzureDevOpsProjectId {
+        param(
+            [string]$Organization,
+            [string]$Project
+        )
+
+        if (!$azureDevOpsProjectIdCache.$Project) {
+            Write-Host "Getting project id for Organization: $Organization project: $Project"
```
Comment (admin-simeon) [inline SimeonInstaller.ps1:604]: no Write-Host (Write-Verbose or Write-Information or WriteWarning)

File: SimeonInstaller.ps1 [hunk 10]
```diff
@@ -1657,6 +1769,358 @@ CRLFOption=CRLFAlways
...
+"@
+                } | Out-Null
+            }
+        }
+
+        ## Project settings
+        Write-Information "Configuring pipeline settings"
+        Invoke-Command -ScriptBlock {
+            # Pipelines > Settings > uncheck Limit job authorization scope to current project for non-release pipelines (enforceReferencedRepoScopedToken), Limit job authorization scope to referenced Azure DevOps repositories (enforceJobAuthScope), and Limit variables that can be set at queue time (enforceSettableVar)
+            Write-Information "Limiting job authorization scope to current project for non-release pipelines"
```
Comment (admin-simeon) [inline SimeonInstaller.ps1:1940]: aren't you doing the OPPOSITE? "uncheck Limit job authorization scope to current project for non-release pipelines "

**Review by admin-simeon — COMMENTED** [review]
Nice code!

File: SimeonInstaller.ps1 [hunk 11]
```diff
@@ -1657,6 +1778,389 @@ CRLFOption=CRLFAlways
...
+        # If org was created now, remove connection to aad and make sure the selection for new url is selected
+        if ($createdOrg) {
+
+            Write-Information "Checking box to use new URL"
+            Invoke-Command -ScriptBlock {
+                Invoke-WithRetry { Invoke-RestMethod -Headers $authenicationHeader -Uri "https://vssps.dev.azure.com/$Organization/_apis/NewDomainUrlOrchestration?codexDomainUrls=true&api-version=5.0-preview.1" -Method Post -ContentType "application/json" }
+            }
+
+            Invoke-Command -ScriptBlock {
+                Write-Verbose "Disconnecting Organization from Azure Active Directory"
```
Comment (admin-simeon) [inline SimeonInstaller.ps1:2146]: why verbose?

File: SimeonInstaller.ps1 [hunk 12]
```diff
@@ -512,6 +515,124 @@ CRLFOption=CRLFAlways
...
+            [ValidateNotNullOrEmpty()]
+            [string]$ProjectId,
+            [ValidateNotNullOrEmpty()]
+            [string]$SubjectGroupPrincipalName,
+            [ValidateNotNullOrEmpty()]
+            [string]$PermissionDescription,
+            [ValidateNotNullOrEmpty()]
+            [int]$PermissionNumber
+        )
+        $groups = (Invoke-WithRetry { Invoke-RestMethod -Header $authenicationHeader -Uri "https://vssps.dev.azure.com/$Organization/_apis/graph/groups?api-version=6.1-preview.1" -Method Get }).value
```
Comment (admin-simeon) [inline SimeonInstaller.ps1:570]: you shouldn't assume this variable is defined in a parent scope - each function should get its own auth header

**Review by admin-simeon — CHANGES_REQUESTED** [review]

**Comment by Lance-Simeon** [comment]
Branch is out of date

**Pull request closed without merging**